src/Embedders/Word2Vec.php

```diff
@@ -0,0 +1,924 @@
+    }
+
+    /**
+     * Embed a high dimensional dataset into a lower dimensional one.
+     *
+     * @param array[] $samples
+     * @throws \RuntimeException
+     */
+    public function transform(array &$samples) : void
+    {
+        if (!$this->vectorsNorm) {
+            throw new RuntimeException('Transformer has not been fitted.');
+        }
+
+        foreach ($samples as &$sample) {
+            foreach ($sample as $i2 => $sentence) {
+                $preppedSentence = $this->prepSentence($sentence);
+
+                $embeddings = [];
+                foreach ($preppedSentence as $word) {
+                    $embeddings[] = $this->embedWord($word);
+                }
+
+                unset($sample[$i2]);
+            }
+
+            if (isset($embeddings)) {
+                $sample = Matrix::stack($embeddings)->transpose()->mean()->asArray();
+            }
+        }
+    }
+
+    /**
+     * Determining the top N similar words provided an array of positive words and negative words.
+     *
+     * @param string[] $positive
+     * @param string[] $negative
+     * @param int $top
+     * @return string[] $result
+     */
+    public function mostSimilar(array $positive, array $negative = [], $top = 20) : array
+    {
+        $positiveArray = $negativeArray = $allWords = $means = [];
+
+        foreach ($positive as $word) {
+            $positiveArray[$word] = 1.0;
+        }
+
+        foreach ($negative as $word) {
+            $negativeArray[$word] = -1.0;
+        }
+
+        $wordArray = array_merge($positiveArray, $negativeArray);
+
+        foreach ($wordArray as $word => $weight) {
+            $wordEmbedding = $this->wordVec($word);
+
+            if ($wordEmbedding instanceof Vector) {
+                $means[] = $wordEmbedding->multiplyScalar($weight);
+                $allWords[] = $this->vocab[$word]['index'];
+            }
+        }
+
+        if (empty($allWords)) {
+            throw new InvalidArgumentException('Positive words were not found in vocab.');
+        }
+
+        $mean = Matrix::stack($means)->transpose()->mean();
+        $l2 = Matrix::stack($this->vectorsNorm);
+        $dists = $mean->transpose()->matmul($l2->transpose())->asArray()[0];
+
+        arsort($dists, 1);
+
+        $result = [];
+        foreach ($dists as $index => $weight) {
+            if (!in_array($index, $allWords)) {
+                $result[$this->index2word[$index]] = $weight;
+            }
+        }
+
+        return array_slice($result, 0, $top, true);
+    }
+
+    /**
+     * Returns the word embedding for a given word.
+     *
+     * @param string $word
+     * @param bool $useNorm
+     * @return \Tensor\Vector|null $result
+     */
+    public function wordVec(string $word, bool $useNorm = true) : ?Vector
+    {
+        if (!array_key_exists($word, $this->vocab)) {
+            return null;
+        }
+
+        if ($useNorm) {
+            $result = $this->vectorsNorm[$this->vocab[$word]['index']];
```

Comment:
Do we need this assignment or can we return right away from within the if statement?